UnicornRender/video/source/Camera.cpp

```diff
@@ -12,120 +12,290 @@ namespace unicorn
 {
-    return m_matrices.m_projection;
+    m_cameraView = glm::lookAt(m_position,
+                               m_position + m_direction,
+                               m_upVector);
 }
 
-const glm::mat4& Camera::GetView() const
+CameraFpsController::CameraFpsController(glm::mat4& cameraView, double posX, double posY)
+    : CameraController(cameraView),
+    sensitivity(0.1f),
+    m_lastX(posX),
+    m_lastY(posY),
+    m_yaw(90.0),
+    m_pitch(0.0)
 {
-    return m_matrices.m_view;
 }
 
-float Camera::GetFov() const
+void CameraFpsController::MoveUp(float deltaTime)
 {
-    return m_fov;
+    m_position += glm::vec3(0.0f, -deltaTime * m_speed, 0.0f);
+    m_isDirty = true;
 }
 
-void Camera::Frame()
+void CameraFpsController::MoveDown(float deltaTime)
 {
-    if (m_dirtyView)
-    {
-        UpdateViewMatrix();
-        m_dirtyView = false;
-    }
-    if (m_dirtyProjection)
+    m_position += glm::vec3(0.0f, m_speed * deltaTime, 0.0);
+    m_isDirty = true;
+}
+
+void CameraFpsController::MoveLeft(float deltaTime)
+{
+    m_position -= m_rightVector * m_speed * deltaTime;
+    m_isDirty = true;
+}
+
+void CameraFpsController::MoveRight(float deltaTime)
+{
+    m_position +=m_rightVector * m_speed * deltaTime;
+    m_isDirty = true;
+}
+
+void CameraFpsController::MoveForward(float deltaTime)
+{
+    m_position += m_direction * m_speed * deltaTime;
+    m_isDirty = true;
+}
+
+void CameraFpsController::MoveBackward(float deltaTime)
+{
+    m_position -= m_direction * m_speed * deltaTime;
+    m_isDirty = true;
+}
+
+void CameraFpsController::UpdateView(double posX, double posY)
+{
+    double xoffset = m_lastX - posX;
+    double yoffset = posY - m_lastY;
+    m_lastX = posX;
+    m_lastY = posY;
+
+    xoffset *= sensitivity;
+    yoffset *= sensitivity;
+
+    m_yaw += xoffset;
+    m_pitch += yoffset;
+
+    m_pitch = std::max(std::min(m_pitch, 89.0), -89.0);
+
+    glm::vec3 front;
+    front.x = static_cast<float>(cos(glm::radians(m_yaw)) * cos(glm::radians(m_pitch)));
+    front.y = static_cast<float>(sin(glm::radians(m_pitch)));
+    front.z = static_cast<float>(sin(glm::radians(m_yaw)) * cos(glm::radians(m_pitch)));
+    m_direction = glm::normalize(front);
+    m_rightVector = glm::normalize(glm::cross(m_direction, m_upVector));
+
+    m_isDirty = true;
+}
+
+UNICORN_EXPORT void CameraFpsController::UpdateViewPositions(double posX, double posY)
+{
+    m_lastX = posX;
+    m_lastY = posY;
+}
+
+CameraProjection::CameraProjection(system::Window* window, glm::mat4& cameraProj) :
+    m_aspect(static_cast<float>(window->GetSize().first) / window->GetSize().second),
+    m_cameraProjection(cameraProj),
+    m_pWindow(window)
+{
+    m_pWindow->Destroyed.connect(this, &CameraProjection::Disconnect);
+    m_pWindow->SizeChanged.connect(this, &CameraProjection::OnWindowSizeChanged);
+}
+
+CameraProjection::~CameraProjection()
+{
+    Disconnect(nullptr);
+}
+
+void CameraProjection::Disconnect(system::Window*)
```

Comment:
if you don't use the argument, why have it at all?

Reply:
Slot prototype for window signal